### ISSUE TITLE
ffmpeg: compilation with fdk-aacv2 broken

### ISSUE THREAD

**Issue by LigH-de** [jb-alvarado/media-autobuild_suite#907]
As reported in https://github.com/nu774/fdkaac/issues/32 — https://github.com/mstorsjo/fdk-aac/commit/b95b15e51d8c692735df4d38c1335efc06aa0443 broke compilation, e.g. by renaming several struct fields (one example: encoderDelay is now nDelay)

```
../src/main.c: In function 'main':
../src/main.c:849:33: error: 'AACENC_InfoStruct {aka struct <anonymous>}' has no member named 'encoderDelay'; did you mean 'nDelay'?
     if (scale_shift && (aacinfo.encoderDelay & 1)) {
                                 ^~~~~~~~~~~~
                                 nDelay
../src/main.c:863:34: error: 'AACENC_InfoStruct {aka struct <anonymous>}' has no member named 'encoderDelay'; did you mean 'nDelay'?
         uint32_t delay = aacinfo.encoderDelay;
                                  ^~~~~~~~~~~~
                                  nDelay
```

[logs.zip](https://0x0.st/svZM.zip)

**Comment by LigH-de** [comment]
@mstorsjo mentioned a [patch for libav](https://git.libav.org/?p=libav.git;a=commitdiff;h=141c960e21d2860e354f9b90df136184dd00a9a8;hp=c8bca9fe466f810fd484e2c6db7ef7bc83b5a943).

**Comment by mstorsjo** [comment]
Yes; in addition to the renamed fields, there might be a need to adjust how the last packets are flushed from the encoder. However if I read https://github.com/nu774/fdkaac/blob/master/src/aacenc.c correctly, this shouldn't need any changes with respect to that.

**Comment by wiiaboo** [comment]
@mstorsjo like this? Looks a bit ugly.
```diff
diff --git a/src/main.c b/src/main.c
index 1eb9a0e..a979e07 100644
--- a/src/main.c
+++ b/src/main.c
@@ -40,6 +40,11 @@
 #include "version.h"
 #include "metadata.h"
 
+#define FDKENC_VER_AT_LEAST(vl0, vl1) \
+    (defined(AACENCODER_LIB_VL0) && \
+        ((AACENCODER_LIB_VL0 > vl0) || \
+         (AACENCODER_LIB_VL0 == vl0 && AACENCODER_LIB_VL1 >= vl1)))
+
 #define PROGNAME "fdkaac"
 
 static volatile int g_interrupted = 0;
@@ -846,7 +851,11 @@ int main(int argc, char **argv)
         m4af_set_priming_mode(m4af, params.gapless_mode + 1);
         m4af_begin_write(m4af);
     }
+#if FDKENC_VER_AT_LEAST(4, 0)
+    if (scale_shift && (aacinfo.nDelay & 1)) {
+#else
     if (scale_shift && (aacinfo.encoderDelay & 1)) {
+#endif
         /*
          * Since odd delay cannot be exactly expressed in downsampled scale,
          * we push one zero frame to the encoder here, to make delay even
@@ -860,7 +869,12 @@ int main(int argc, char **argv)
     if (frame_count < 0)
         goto END;
     if (m4af) {
-        uint32_t delay = aacinfo.encoderDelay;
+        uint32_t delay;
+#if FDKENC_VER_AT_LEAST(4, 0)
+        delay = aacinfo.nDelay;
+#else
+        delay = aacinfo.encoderDelay;
+#endif
         uint32_t padding;
         int64_t frames_read = pcm_get_position(reader);
 
```

**Comment by mstorsjo** [comment]
Yes, that looks correct, but yes, also ugly. (You could play with something like `#if FDKENC_VER_AT_LEAST(4, 0) #define encoderDelay nDelay #endif` but that's also a potential recipe for a much bigger mess later on.)

**Comment by LigH-de** [comment]
https://github.com/nu774/fdkaac was updated, I'll test again ...

**Comment by wiiaboo** [comment]
Upstream fixed it.

**Comment by schmidthubert** [comment]
Is it working now or is it me? I have (after a complete cleanup of everything *fdk*):
`me@host /trunk/local64/include`
`$ grep -inr AACENCODER_LIB_VL0 *`
`fdk-aac/aacenc_lib.h:1028:#define AACENCODER_LIB_VL0 4`

but I'm still getting the ugly

`has no member named 'encoderDelay'; did you mean 'nDelay'?`

error? I checked my ORIG_HEADS to be sure and they are:
`651ff34d8d35fb6a3b75471d54b271852f5924cc`
for fdk-aac.git
and
`f2f9c1b4bbc0c6c579dc79a2ce5277d2dc505551`
for fdkaac.git

so I'm up-to-date. Still no dice ?

**Comment by LigH-de** [comment]
I just had a compile in MinGW32 passing fdk-aac and bin-fdk-aac, got a fresh local32/bin-audio/fdkaac.exe (before it broke elsewhere).

**Comment by schmidthubert** [comment]
It's not that, sorry I didn't say where I get this. I have that exe as well and both routines go through the build just fine. It is make ffmpeg which is throwing that error now ...

**Comment by LigH-de** [comment]
Commenting out liblensfun to test that, one moment...

**Comment by schmidthubert** [comment]
OK, thanks! Will check again then where this is coming from, very strange. Will clean out even more then :+1:

**Comment by LigH-de** [comment]
I get even slightly different behaviours per PC I'm running it on. I guess it's time for me to clean up deeply.

**Comment by schmidthubert** [comment]
That info.encoderDelay error comes from
`trunk/build/ffmpeg-git/libavcodec/libfdk-aacenc.c:292:    avctx->initial_padding = info.encoderDelay;`
It can be patched away the same way the fdkaac team did. But apparently dozens of ffmpegs devs aren't even using fdk-aac, nor do they do any tests of ALL of their code (otherwise they would have noticed even under linux) or everyone is sleeping as usual ... Whatever, my fdk build works again :+1: 
![fdk](https://user-images.githubusercontent.com/26191502/45090445-a4a98900-b10f-11e8-8cda-7431370c04d8.PNG)

**Comment by LigH-de** [comment]
Regarding liblensfun and missing gnulib, I reported to https://github.com/Alexpux/MINGW-packages/issues/4372

**Comment by wiiaboo** [comment]
fdk-aac FFmpeg-side is broken because @mstorsjo's patch is for libav, don't remember if he also submitted it to ffmpeg-devel.

**Comment by schmidthubert** [comment]
I've checked with ffmpeg pipermail devel and patchwork before I posted this and there's zilch about this. It is becoming more and more of a habit these days writing code but don't test at all. Look at M$ at the update shit they produce weekly, and that's just another example ...

**Comment by wiiaboo** [comment]
He just didn't submit it this time. There's no new code being written to be tested, that's the issue.
Unless you were expecting Google to test mstorjo's port to libfdkaac working in everything depending on it.

**Comment by schmidthubert** [comment]
No, not mstorsjo needs to test. But ffmpeg devs need to compile their code sometimes with ALL options they SAY they support!

**Comment by LigH-de** [comment]
Well, they have me. ¯\_(ツ)_/¯

**Comment by wiiaboo** [comment]
FFmpeg devs don't have to pay attention to every dependent external lib working or breaking. They only have to pay attention to internal code.

**Comment by schmidthubert** [comment]
Yeah, the attitude of most people today turns into "only have to do the necessary but not more", "why me", "what do I care" etc so don't do anything "extra". It is this elite behavior which becomes more and more common. That's why we are where we are today... Totally unrelated to GH, ffmpeg, mabs or anything. Just saying!

**Comment by wiiaboo** [comment]
Libs devs are the ones that should take care not to break too much API, usually. If you make a public lib, you're responsible for it, and if you do introduce breaking changes, they should be clearly noted or communicated as such, afaik.

**Comment by wiiaboo** [comment]
Waiting on https://git.libav.org/?p=libav.git;a=commit;h=141c960e21d2860e354f9b90df136184dd00a9a8 being merged into FFmpeg.

**Comment by wiiaboo** [comment]
Merged upstream.

**Comment by Knot83** [comment]
Hello Goog morning,

I'm getting the following error while compiling ffmpeg following this article (https://gist.github.com/adddog/d82581faf10d666d35b9771eaa65d5ac) : 

`libavcodec/libfdk-aacenc.c:289:34: error: ‘AACENC_InfoStruct {aka struct <anonymous>}’ has no member named ‘encoderDelay’; did you mean ‘nDelay’?`

I've read the previous post. What I did was to copy the content from [this ](https://git.libav.org/?p=libav.git;a=commitdiff;h=141c960e21d2860e354f9b90df136184dd00a9a8;hp=c8bca9fe466f810fd484e2c6db7ef7bc83b5a943)website to a brand new libfdk-aacenc.c and substituted it on my working dir where I'm compiling but it didn't solve the problem. Am I doing a wrong procedure?

Can you please help me to understand how to solve this compiling issue?

Thanks

**Comment by LigH-de** [comment]
Hi.

I see no reason to apply some complicated manual patching to compile ffmpeg including fdk-aac when you want to use the media-autobuild suite. You will only need to both set up the suite to build a "Non-free" edition to allow mixed licenses (then you have to keep the result private) and enable building of fdk-aac. If you also select to build ffmpeg with optional libraries, you can edit a text file and include or exclude the libraries you prefer.

**Comment by Knot83** [comment]
Hi! Thanks!
I'm sorry but I'd appreciate if you could help me detailing the procedure you're referring to. I'm stack in row 50 of the [procedure ](https://gist.github.com/adddog/d82581faf10d666d35b9771eaa65d5ac)unable to complete the ffmpeg installation. Believe me I've followed several procedure but it seems I need someone that can follow me in the procedure.
Without ffmpeg I cannot use the Stream component of Home Assistant and consequently my IPCAM, please help me.

p.s. I see media-autobuild suite is for Windows. I'm running Raspbian on Raspberry.

Thanks

**Comment by LigH-de** [comment]
In this case, I would expect wiiaboo telling you just the same as already explained in the ReadMe chapter of the project page: If it is not a matter specific to the workflow of this suite, this issue tracker is probably not the perfect place to discuss it. I could suggest you an ffmpeg mailing list, or the hydrogenaud.io forums (a home of the author of the fdk-aac build suite which was integrated into MABS later).

**Comment by Knot83** [comment]
Do you have any working procedure to install FFMPEG in my case?

**Comment by Knot83** [comment]
In the meantime I wrote this post https://hydrogenaud.io/index.php/topic,118111.new.html#new. Hoping someone can help me.

**Comment by wiiaboo** [comment]
This is an issue that should be reported to FFmpeg, not fixed by us or by applying a patch from libav.
